keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -242,22 +242,44 @@ type AdmissionRequest struct {
 
 ### Mutating Plugin ordering
 
-Current ordering of the plugins (including webhooks) is fixed but may not work for all cases 
-(e.g. [this issue](https://github.com/kubernetes/kubernetes/issues/64333)). A mutating webhook 
-can add a completely new structure to the object (e.g. a `Container`) and other mutating plugins 
-may have opinion on those new structures (e.g. setting the image pull policy on all containers). 
-Although this problem can go deeper than two level (if the second mutating webhook added a new 
-structure based on the new `Container`), with current set of standard mutating plugins and use-cases, 
-it look like a rerun of mutating plugins is a sufficient fix. The proposal is to rerun all mutating 
-plugins (including webhooks) if there is any change by mutating webhooks. The assumption for this 
-process is all mutating plugins are idempotent and this must be clearly documented to users.
+A single ordering of mutating admissions plugins (including webhooks) does not work for all cases
+(see https://issue.k8s.io/64333 as an example). A mutating webhook can add a new sub-structure 
+to the object (like adding a `container` to a pod), and other mutating plugins which have already 
+run may have opinions on those new structures (like setting an `imagePullPolicy` on all containers). 
+
+To allow mutating admission plugins to observe changes made by other plugins, regardless of order,
+admission will add the ability to re-run mutating plugins (including webhooks) a single time if 
+there is any change made by a mutating webhook on the first pass.
+
+1. Initial mutating admission pass
+2. If any webhook plugins returned patches modifying the object, do a second mutating admission pass
+  1. Run all in-tree mutating admission plugins
+  2. Run all applicable webhook mutating admission plugins that indicate they should be re-called
+
+Mutating plugins that are re-called must be able to successfully admit an object they have already 
+admitted and potentially modified. This will be clearly documented in admission webhook documentation,
+examples, and test guidelines. Mutating admission webhooks *should* already support this (since any 
+change they can make in an object could already exist in the user-provided object), and any webhooks
+that do not are broken for some set of user input, but for compatibility, the re-call behavior will be 
+opt-in for `v1beta1`. In `v1`, mutating webhooks must be safe to recall.
 
-Mutations from in-tree plugins will not trigger this process as they are well-ordered but if 
-there is any mutation by webhooks, all of the plugins including in-tree ones will be run again.
-
-This feature would be would be opt in and defaulted to false for `v1beta1`.
+```golang
+type MutatingWebhookConfiguration struct {
+    ...
+    // unsafeToRecall indicates whether these mutating webhooks can be called again 
```

Comment:
Current name sounds like a plot element from "Inception" or "Minority Report" :)

Ideas:

1. `Idempotent`: default false in v1beta1; in v1, webhooks must be both idempotent and side-effect free? Comments need to explain the difference.
2. `MultiInvocationSafe`? I think it's better to state in the positive, and I think "recall" doesn't parse well.
3. `WantsMultipleInvocations`: I think it's quite plausible that many webhooks just need to do what they do without regard to anything else, e.g. a webhook that just adds a label or annotation. We don't need to call them multiple times.

Reply:
We were looking for a name that indicates setting this to false is bad and you should feel bad about setting it.  There are some cases where you could optimize away the re-call, but without a strong, specific need for the performance boost, having a single set of rules makes it easier to doc and eliminates the ability for a user to mess up.

Reply:
renamed field to `idempotent`. If in the future we want to optimize opting out/in of re-call, I think that should be dealt with separately, and probably made part of the admissionreview response rather than the webhook registration config, so I'm in favor of limiting this to v1beta1 to avoid breaking currently buggy webhooks, and start v1 without the config knob.

Reply:
I feel the opposite way-- wanting multiple invocations is bad and you should feel bad for making a webhook with a weird dependency chain.

I also think 30 is not an implausible number of webhooks in the nearish future. Doubling that to 60 calls is much worse (e.g. for tail latency) than e.g. increasing to 40.

Reply:
> you should feel bad for making a webhook with a weird dependency chain.

one webhook injecting a sidecar container and another injecting envvars to all containers seems like a very reasonable scenario where the two webhooks should not need knowledge of each other

Reply:
you're saying the webhook that injects envvars to all containers should explicitly opt into being recalled, and the one that injects a new container should not? I'm not terribly opposed to a long-term opt-in mechanism, but I do think webhook authors are likely to not set it correctly

Reply:
Yeah, exactly. And in this light, I regret my earlier suggestion of "idempotent" :)

I'm not saying the situation is avoidable, but it is still something to be minimized.

I guess I suggest default "no recall" in v1beta1 and maybe just force the user to choose in v1 (required, no default). If it is set wrong it is easy to fix.

Reply:
updated